File: src/main/java/planstack/constraints/bindings/ValuesHolder.java
package planstack.constraints.bindings;


import scala.collection.JavaConversions;
import scala.collection.immutable.BitSet;
import scala.collection.immutable.BitSet$;

import java.util.Collection;
import java.util.Set;

/**
 * Holds values for CSP.
 *
 * This implementation is immutable: it won't get any change once the constructor is called.
 * Sca immutable BitSet is used as internal representation, hence the domain
 * <code>{0, 2, 3, 7}</code> is represented by the BitSet <code>10110001</code>.
 *
 * Any modification to a ValuesHolder returns a new ValuesHolder object with an updated BitSet.
 */
public class ValuesHolder {
    BitSet values;

    ValuesHolder(Collection<Integer> values) {
        this.values = BitSet$.MODULE$.empty();
        for(int val : values) {
            this.values = this.values.$plus(val);
        }
    }

    ValuesHolder(BitSet values) {
        this.values = values;
    }

    public boolean isEmpty() {
        return values.isEmpty();
    }

    public boolean equals(ValuesHolder o) {
        return this == o || this.values.equals(o.values);
    }

    public ValuesHolder intersect(ValuesHolder holder) {
        return new ValuesHolder((BitSet) this.values.$amp(holder.values));
    }

    /**
     * @return True if the at least one of the given values is present in this domain.
     */
    public boolean containsAtLeastOne(Collection<Integer> possibleValues) {
        for(int val : possibleValues) {
            if(values.contains(val))
                return true;
        }
        return false;
    }

    public Set<Integer> values() {
        return (Set) JavaConversions.setAsJavaSet(values);
    }

    public int size() {
        return values.size();
    }

    public ValuesHolder remove(int value) {
        return new ValuesHolder(values.$minus(value));
    }

    public ValuesHolder remove(ValuesHolder toRemove) {
        return new ValuesHolder((BitSet) values.$amp$tilde(toRemove.values));
    }

    public boolean contains(int value) {
        return values.contains(value);
    }
}

File: src/main/scala/planstack/constraints/bindings/ValuesHolder.scala
package planstack.constraints.bindings

import java.util

import scala.collection.JavaConversions
import scala.collection.JavaConverters._
import scala.collection.immutable.BitSet

object ValuesHolder {
  val BITSET_MAX_SIZE = 1000
  def convert(values: util.Collection[Integer]) : Set[Int] = {
    var max = -10
    for(v <- values.asScala)
      max = if(v > max) v else max
    var set : Set[Int] =
      if(max > BITSET_MAX_SIZE)
        Set[Int]()
      else
        BitSet()
    set ++= values.asInstanceOf[util.Collection[Int]].asScala
    set
  }
}

class ValuesHolder(val vals: scala.collection.Set[Int]) {

  def this(values: util.Collection[Integer]) = this(ValuesHolder.convert(values))
  def this(values: Iterable[Int]) = this(values.map(_.asInstanceOf[Integer]).asJavaCollection)

  lazy private val _isEmpty = vals.isEmpty
  lazy private val _size = vals.size

  def values() : util.Set[Integer] = JavaConversions.setAsJavaSet(vals).asInstanceOf[java.util.Set[Integer]]

  def size() : Integer = _size

  def head() : Integer = vals.head

  def intersect(other: ValuesHolder) : ValuesHolder = {
    val intersection = (vals, other.vals) match {
      case (v1: BitSet, v2:BitSet) => v1 & v2 // should be significantly faster as it is just and 'and' on two bitset
      case (v1, v2) => v1 & v2
    }
    new ValuesHolder(intersection)
  }

  def union(other: ValuesHolder) : ValuesHolder = {
    val union = (vals, other.vals) match {
      case (v1: BitSet, v2: BitSet) => v1 | v2 // should be significantly faster as it is just and 'or' on two bitset
      case (v1, v2) => v1 | v2
    }
    new ValuesHolder(union)
  }
  def contains(v: Integer) : Boolean =
    vals.contains(v)

  def isEmpty : Boolean = _isEmpty

  def nonEmpty = !isEmpty

  def remove(toRm: ValuesHolder) : ValuesHolder =
    new ValuesHolder(vals -- toRm.vals)

  def remove(toRm: Integer) : ValuesHolder =
    new ValuesHolder(vals - toRm)

  def add(value: Integer) : ValuesHolder = {
    if (vals.isInstanceOf[BitSet] && value > ValuesHolder.BITSET_MAX_SIZE)
    // it has grown too big, switch to too normal Set
      new ValuesHolder(Set[Int]() ++ vals + value)
    else
      new ValuesHolder(vals + value)
  }
}
